### PR TITLE
Using SSA with moco-controller

### PR DESCRIPTION
**Pull request by d-kuro**
refs: https://github.com/cybozu-go/moco/issues/306

Using SSA with moco-controller.

## Points

* Embedded applyconfigurations in CRD of `MySQLCluster`.
  * This will allow to include a description of the CRD: https://github.com/cybozu-go/moco/pull/364/commits/639d203f3568d46b4847a395785cb021c96e2c4d
  * As a disadvantage, some CRD validate information has been lost.
    * The applyconfigurations struct does not contain any comments for validation.
* Currently,  using SSA only for creating/editing Service and StatefulSet.
  * The difference will be huge, so I will split the PullRequest for the rest.

### CODE REVIEW

File: api/v1beta1/conversion_test.go [hunk 1]
```diff
@@ -15,7 +17,14 @@ func TestCompatibility(t *testing.T) {
...
 
-	f := roundtrip.CompatibilityTestFuzzer(scheme, nil)
+	// Suppress typed nil.
+	fn := func(l **corev1.ResourceList, c fuzz.Continue) {
+		if l != nil && *l == nil || **l == nil {
+			*l = nil
+		}
+	}
+
+	f := roundtrip.CompatibilityTestFuzzer(scheme, []interface{}{fn})
```
Comment (d-kuro) [inline api/v1beta1/conversion_test.go:27]: Needed to suppress typed nil.

```
    conversion_test.go:53: compatibility error case #0 (-want +got):
          v1beta1.MySQLCluster{
          	TypeMeta:   {},
          	ObjectMeta: {Name: "2", GenerateName: "3", Namespace: "4", SelfLink: "5", ...},
          	Spec: v1beta1.MySQLClusterSpec{
          		Replicas:    1971674115,
          		PodTemplate: {ObjectMeta: {Name: "24", Labels: {"25": "26", "27": "28", "29": "30"}}, Spec: {EphemeralContainers: {{EphemeralContainerCommonApplyConfiguration: {Name: &"31", Image: &"32", Command: {"33", "34"}, Args: {"35"}, ...}, TargetContainerName: &"74"}}, TerminationGracePeriodSeconds: &1409060769374878694, NodeSelector: {"75": "76"}, ServiceAccountName: &"77", ...}},
          		VolumeClaimTemplates: []v1beta1.PersistentVolumeClaim{
          			{
          				ObjectMeta: {Name: "89", Annotations: {"90": "91", "92": "93", "94": "95"}},
          				Spec: v1beta1.PersistentVolumeClaimSpecApplyConfiguration{
          					AccessModes: {},
          					Selector:    nil,
          					Resources: &v1.ResourceRequirementsApplyConfiguration{
          						Limits:   &{s"Ŗb$": {i: {...}, Format: "DecimalExponent"}},
        - 						Requests: &nil,
        + 						Requests: nil,
          					},
          					VolumeName:       &"96",
          					StorageClassName: &"97",
          					... // 3 identical fields
          				},
          			},
```

File: Makefile [hunk 2]
```diff
@@ -28,7 +28,7 @@ SHELL = /bin/bash
 .SHELLFLAGS = -e -o pipefail -c
 
 # Produce CRDs that work back to Kubernetes 1.11 (no version conversion)
-CRD_OPTIONS = "crd:crdVersions=v1,maxDescLen=0"
+CRD_OPTIONS = "crd:crdVersions=v1,maxDescLen=105"
```
Comment (d-kuro) [inline Makefile:31]: In the current CRD, `maxDescLen=109` was the limit.

File: docs/crd_mysqlcluster_v1beta2.md [hunk 3]
```diff
@@ -136,7 +136,7 @@ PersistentVolumeClaim is a user's request for and claim to a persistent volume.
 | Field | Description | Scheme | Required |
 | ----- | ----------- | ------ | -------- |
 | metadata | Standard object's metadata. | [ObjectMeta](#objectmeta) | true |
-| spec | Spec defines the desired characteristics of a volume requested by a pod author. | [corev1.PersistentVolumeClaimSpec](https://pkg.go.dev/k8s.io/api/core/v1#PersistentVolumeClaimSpec) | true |
+| spec | Spec defines the desired characteristics of a volume requested by a pod author. | PersistentVolumeClaimSpecApplyConfiguration | true |
```
Comment (ymmt2005) [inline docs/crd_mysqlcluster_v1beta2.md:139]: Please update https://github.com/cybozu-go/moco/blob/main/docs/links.csv to link to the appropriate API manual.
In this case, it is https://pkg.go.dev/k8s.io/client-go@v0.23.3/applyconfigurations/core/v1#PersistentVolumeClaimApplyConfiguration

**Review by ymmt2005 — CHANGES_REQUESTED** [review]

**Review by masa213f — CHANGES_REQUESTED** [review]

File: controllers/mysql_container.go [hunk 4]
```diff
@@ -6,287 +6,275 @@ import (
...
-	var containers []corev1.Container
-	for _, c := range cluster.Spec.PodTemplate.Spec.Containers {
-		switch c.Name {
+func (r *MySQLClusterReconciler) makeV1OptionalContainers(cluster *mocov1beta2.MySQLCluster) []*corev1ac.ContainerApplyConfiguration {
+	var containers []*corev1ac.ContainerApplyConfiguration
+
+	spec := cluster.Spec.PodTemplate.Spec.DeepCopy()
+	for _, c := range spec.Containers {
+		c := c
+		switch *c.Name {
```
Comment (masa213f) [inline controllers/mysql_container.go:225]: Please add a nil check. The moco-controller will be crashed by the following manifest. 

```yaml
apiVersion: moco.cybozu.com/v1beta2
kind: MySQLCluster
metadata:
  name: foo
spec:
  podTemplate:
    spec:
      containers:
      - image: quay.io/cybozu/mysql:8.0.27
        name: mysqld
      - image: invalid      # nameがなく不正なコンテナ定義
  volumeClaimTemplates:
  - metadata:
      name: mysql-data
    spec:
      accessModes: [ "ReadWriteOnce" ]
      resources:
        requests:
          storage: 1Gi
  logRotationSchedule: "@every 10s"
```

File: controllers/mysqlcluster_controller.go [hunk 5]
```diff
@@ -498,254 +514,269 @@ func (r *MySQLClusterReconciler) reconcileV1Service(ctx context.Context, req ctr
...
+		containers = append(containers, r.makeV1SlowQueryLogContainer(sts, force))
+	}
+	if len(cluster.Spec.Collectors) > 0 {
+		containers = append(containers, r.makeV1ExporterContainer(cluster.Spec.Collectors))
+	}
+	containers = append(containers, r.makeV1OptionalContainers(cluster)...)
 
-		podSpec := cluster.Spec.PodTemplate.Spec.DeepCopy()
-		podSpec.ServiceAccountName = cluster.PrefixedName()
+	initContainers := r.makeV1InitContainer(cluster, *mysqldContainer.Image)
```
Comment (masa213f) [inline controllers/mysqlcluster_controller.go:729]: Please add nil check or validation for the `image` field of `mysqld` container. 

```
apiVersion: moco.cybozu.com/v1beta2
kind: MySQLCluster
metadata:
  name: foo
spec:
  podTemplate:
    spec:
      containers:
      - name: mysqld
  volumeClaimTemplates:
  - metadata:
      name: mysql-data
    spec:
      accessModes: [ "ReadWriteOnce" ]
      resources:
        requests:
          storage: 1Gi
```

**Review by masa213f — CHANGES_REQUESTED** [review]

File: controllers/mysqlcluster_controller.go [hunk 6]
```diff
@@ -498,254 +514,269 @@ func (r *MySQLClusterReconciler) reconcileV1Service(ctx context.Context, req ctr
...
+
+	obj, err := runtime.DefaultUnstructuredConverter.ToUnstructured(svc)
+	if err != nil {
+		return fmt.Errorf("failed to convert Service %s/%s to unstructured: %w", cluster.Namespace, name, err)
+	}
+	patch := &unstructured.Unstructured{
+		Object: obj,
+	}
+
+	var orig, updated corev1.Service
```
[the remark this answers is not in the document]
Comment (d-kuro) [reply controllers/mysqlcluster_controller.go]: Thanks! I moved var definition.
https://github.com/cybozu-go/moco/pull/364/commits/405aa42a198c24308ba96accdabed735cbf829b1

File: controllers/mysqlcluster_controller.go (hunk 5 above)
Comment (d-kuro) [reply controllers/mysqlcluster_controller.go:734]: Fixed: https://github.com/cybozu-go/moco/pull/364/commits/ffce0d95e591711bc73051c70f11b8b6e2904561

File: controllers/mysqlcluster_controller.go [hunk 7]
```diff
@@ -498,254 +514,269 @@ func (r *MySQLClusterReconciler) reconcileV1Service(ctx context.Context, req ctr
...
+		if p.Name == nil {
+			continue
 		}
 
-		saSpec.DeepCopyInto(&svc.Spec)
-		if debugController {
-			updated = svc.Spec.DeepCopy()
+		switch *p.Name {
+		case constants.MySQLPortName:
+			mysqlNodePort = *p.NodePort
```
[the remark this answers is not in the document]
Comment (d-kuro) [reply controllers/mysqlcluster_controller.go:550]: Fixed: https://github.com/cybozu-go/moco/pull/364/commits/ffce0d95e591711bc73051c70f11b8b6e2904561

File: docs/crd_mysqlcluster_v1beta2.md (hunk 3 above)
Comment (d-kuro) [reply docs/crd_mysqlcluster_v1beta2.md]: Thanks you, I fixed:
https://github.com/cybozu-go/moco/pull/364/commits/41fa72ed465531c07620e198f7e6a49e994473c4

**Review by ymmt2005 — APPROVED** [review]
LGTM

**Review by masa213f — APPROVED** [review]
LGTM. Thank you!

**Pull request merged**